app/controllers/users/confirmations_controller.rb

```diff
@@ -60,33 +66,44 @@ def process_user_with_password_errors
     end
 
     def process_user_with_confirmation_errors
+      return process_already_confirmed_user if @confirmable.confirmed?
+
       set_view_variables
       render :new
     end
 
+    def process_already_confirmed_user
+      action_text = 'Please sign in.' unless user_signed_in?
+      flash[:error] = t('devise.confirmations.already_confirmed', action: action_text)
+
+      redirect_to user_signed_in? ? dashboard_index_url : new_user_session_url
+    end
+
     def process_unconfirmed_user
       set_view_variables
 
       if resource.confirmation_period_expired?
         flash[:error] = UserDecorator.new(resource).confirmation_period_expired_error
```

Comment:
You are right. In the case below, where I changed the flash to #now, the user clicks on the link in the email, then ends up on ConfirmationsController#show with a flash message saying their email was successfully confirmed, but that they need to set a password to continue creating their account. After a user enters a valid password, they get redirected to the 2FA setup page. If we don't use `#now` for the flash message, it will show up on the 2FA setup page, which is not what we want.

As for the flash error you commented about here, it appears when the user takes too long to click the link in the email. They end up on a screen that asks them to send confirmation instructions again. When they click the button, they get redirected to the home page with a new flash message, so the old one won't get carried over. However, if they decide to navigate away instead of resending instructions, by going to the home page, for example, the message will carry over. Since they can only go to either the home page or the sign up page, perhaps it makes sense to keep this message around in this case to let them know they still need to confirm their email?